### ISSUE TITLE
soft placement not working

### ISSUE THREAD

**Issue by kerokeroloid** [tensorflow/tensorflow#636]
I have this code:

```
    sess = tf.Session(config=tf.ConfigProto(allow_soft_placement=True,
                                        log_device_placement=True))
    with tf.device("/gpu:0"):
        s, readout, h_fc1 = createNetwork()
    with sess:
            trainNetwork(s, readout, h_fc1, sess)
```

This fails with this error:

```
W tensorflow/core/common_runtime/executor.cc:1076] 0x28d7890 Compute status: Unimplemented: Depthwise max pooling is currently only implemented for CPU devices.
	 [[Node: MaxPool = MaxPool[ksize=[1, 1, 1, 512], padding="SAME", strides=[1, 1, 1, 512], _device="/job:localhost/replica:0/task:0/gpu:0"](Relu_2)]]
W tensorflow/core/common_runtime/executor.cc:1076] 0x2a693f0 Compute status: Unimplemented: Depthwise max pooling is currently only implemented for CPU devices.
	 [[Node: MaxPool = MaxPool[ksize=[1, 1, 1, 512], padding="SAME", strides=[1, 1, 1, 512], _device="/job:localhost/replica:0/task:0/gpu:0"](Relu_2)]]
	 [[Node: add_4/_3 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_1039_add_4", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]
Traceback (most recent call last):
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/client/session.py", line 428, in _do_run
    target_list)
tensorflow.python.pywrap_tensorflow.StatusNotOK: Unimplemented: Depthwise max pooling is currently only implemented for CPU devices.
	 [[Node: MaxPool = MaxPool[ksize=[1, 1, 1, 512], padding="SAME", strides=[1, 1, 1, 512], _device="/job:localhost/replica:0/task:0/gpu:0"](Relu_2)]]
	 [[Node: add_4/_3 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_1039_add_4", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "bob_learning.py", line 222, in <module>
    main()
  File "bob_learning.py", line 219, in main
    playGame()
  File "bob_learning.py", line 216, in playGame
    trainNetwork(s, readout, h_fc1, sess)
  File "bob_learning.py", line 141, in trainNetwork
    readout_t = readout.eval(feed_dict = {s : [s_t]})[0]
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/framework/ops.py", line 460, in eval
    return _eval_using_default_session(self, feed_dict, self.graph, session)
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/framework/ops.py", line 2910, in _eval_using_default_session
    return session.run(tensors, feed_dict)
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/client/session.py", line 368, in run
    results = self._do_run(target_list, unique_fetch_targets, feed_dict_string)
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/client/session.py", line 444, in _do_run
    e.code)
tensorflow.python.framework.errors.UnimplementedError: Depthwise max pooling is currently only implemented for CPU devices.
	 [[Node: MaxPool = MaxPool[ksize=[1, 1, 1, 512], padding="SAME", strides=[1, 1, 1, 512], _device="/job:localhost/replica:0/task:0/gpu:0"](Relu_2)]]
	 [[Node: add_4/_3 = _Recv[client_terminated=false, recv_device="/job:localhost/replica:0/task:0/cpu:0", send_device="/job:localhost/replica:0/task:0/gpu:0", send_device_incarnation=1, tensor_name="edge_1039_add_4", tensor_type=DT_FLOAT, _device="/job:localhost/replica:0/task:0/cpu:0"]()]]
Caused by op 'MaxPool', defined at:
  File "bob_learning.py", line 222, in <module>
    main()
  File "bob_learning.py", line 219, in main
    playGame()
  File "bob_learning.py", line 214, in playGame
    s, readout, h_fc1 = createNetwork()
  File "bob_learning.py", line 98, in createNetwork
    h_pool3_flat = tf.reshape(max_pool_featurewise(h_conv3, old_num_filters), [-1, IMSIZE * IMSIZE])
  File "bob_learning.py", line 52, in max_pool_featurewise
    return tf.nn.max_pool(x, ksize = [1, 1, 1, num_filters], strides = [1, 1, 1, num_filters], padding = "SAME")
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/ops/nn_ops.py", line 235, in max_pool
    name=name)
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/ops/gen_nn_ops.py", line 449, in _max_pool
    strides=strides, padding=padding, name=name)
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/ops/op_def_library.py", line 664, in apply_op
    op_def=op_def)
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/framework/ops.py", line 1834, in create_op
    original_op=self._default_original_op, op_def=op_def)
  File "/mnt/data1/PREZ/PROGRAMMING/2015/12/BOB_learning_0/gpu/lib/python3.4/site-packages/tensorflow/python/framework/ops.py", line 1043, in __init__
    self._traceback = _extract_stack()
```

As far as I understand, it should just place that single op on the CPU and everything else on the GPU.

**Comment by mrry** [comment]
This is a subtly different issue: as far as the placer is concerned MaxPool is defined to run on GPU, so it will be placed on a GPU device if available (using hard or soft placement). However, for certain input values (in this case, where the kernel size in the depth dimension is greater than 1) it will raise an error at runtime if the op tries to execute on GPU.

**TL;DR:** For now it's necessary to annotate any depthwise max-pooling ops with a `with tf.device("/cpu:0"):` block.

**Comment by vrv** [comment]
Yes, this is more about an unsupported configuration of an op, not soft placement.  Retitling this.

**Comment by siddharth-agrawal** [comment]
@vrv I looked into this issue, as far as I can tell adding a `DepthwiseMaxPool` [here](https://github.com/tensorflow/tensorflow/blob/master/third_party/eigen3/unsupported/Eigen/CXX11/src/NeuralNetworks/Pooling.h) will solve the issue. Am I on the right track?

**Comment by vrv** [comment]
We have to remove those header files actually, we moved them into core/kernels/eigen_*

But yes, that's probably where you would have to implement it (thankfully, just once for both CPU and GPU).  I don't know if there's a way to get MaxPool for CuDNN to do depth-pooling, but if so, that should be part of this.

**Comment by aselle** [comment]
@kerokeroloid, is this still an issue for you, or could I close this?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
Nagging Awaiting Response: It has been 14 days with no activityand the `awaiting response` label was assigned. Is this still an issue?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?
